File: Days/1.py
def solve_part1():
    # Read the input file and split into two lists
    with open('Data/1.txt') as f:
        left_nums = []
        right_nums = []
        for line in f:
            left, right = map(int, line.strip().split())
            left_nums.append(left)
            right_nums.append(right)
    
    # Sort both lists
    left_nums.sort()
    right_nums.sort()
    
    # Calculate total distance
    total_distance = sum(abs(l - r) for l, r in zip(left_nums, right_nums))
    
    return total_distance

def solve_part2():
    # Read the input file and split into two lists
    with open('Data/1.txt') as f:
        left_nums = []
        right_nums = []
        for line in f:
            left, right = map(int, line.strip().split())
            left_nums.append(left)
            right_nums.append(right)
    
    # Count occurrences in right list
    right_counts = {}
    for num in right_nums:
        right_counts[num] = right_counts.get(num, 0) + 1
    
    # Calculate similarity score
    total_score = sum(num * right_counts.get(num, 0) for num in left_nums)
    
    return total_score

print(solve_part1())
print(solve_part2())

File: Days/1.rs
use std::error::Error;

type Result<T> = std::result::Result<T, Box<dyn Error>>;

pub fn solve(input: &str) -> Result<(String, String)> {
    let (left, right) = parse_input(input)?;
    
    let part1 = solve_part1(&left, &right)?;
    let part2 = solve_part2(&left, &right)?;
    
    Ok((part1.to_string(), part2.to_string()))
}

fn parse_input(input: &str) -> Result<(Vec<i64>, Vec<i64>)> {
    let mut left = Vec::new();
    let mut right = Vec::new();

    for line in input.lines() {
        if line.trim().is_empty() { continue; }
        
        let mut parts = line.split_whitespace();
        if let (Some(l), Some(r)) = (parts.next(), parts.next()) {
            left.push(l.parse::<i64>()?);
            right.push(r.parse::<i64>()?);
        }
    }

    Ok((left, right))
}

fn solve_part1(left: &[i64], right: &[i64]) -> Result<i64> {
    let mut left_sorted = left.to_vec();
    let mut right_sorted = right.to_vec();
    left_sorted.sort_unstable();
    right_sorted.sort_unstable();

    let total_distance: i64 = left_sorted.iter()
        .zip(right_sorted.iter())
        .map(|(a, b)| (a - b).abs())
        .sum();

    Ok(total_distance)
}

fn solve_part2(left: &[i64], right: &[i64]) -> Result<i64> {
    let similarity_score: i64 = left.iter()
        .map(|&num| {
            let count = right.iter().filter(|&&x| x == num).count() as i64;
            num * count
        })
        .sum();

    Ok(similarity_score)
}

#[cfg(test)]
mod tests {
    use super::*;
    
    const EXAMPLE: &str = r#"3   4
4   3
2   5
1   3
3   9
3   3"#;

    #[test]
    fn test_part1() {
        let (left, right) = parse_input(EXAMPLE).unwrap();
        assert_eq!(solve_part1(&left, &right).unwrap(), 11);
    }

    #[test]
    fn test_part2() {
        let (left, right) = parse_input(EXAMPLE).unwrap();
        assert_eq!(solve_part2(&left, &right).unwrap(), 31);
    }
}
